Makefile

```diff
@@ -90,6 +110,75 @@ test-silent:
 benchmark-test:
 	. ./scripts/shared_env && go test -run=XX -bench=. $(shell go list ./agent/... | grep -v /vendor/)
 
+
+
+
+define dockerbuild
+	docker run \
+		--net none \
+		--user "$(USERID)" \
+		--volume "$(PWD)/out:/out" \
+		--volume "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
+		--rm \
+		$(BUILDER_IMAGE) \
+		$(1)
+endef
+
+define win-cgo-dockerbuild
+	docker run --net=none \
+		--user "$(USERID)" \
+		--volume "$(PWD)/out:/out" \
+		--volume "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
+		--env "GOOS=windows" \
+		--env "CGO_ENABLED=1" \
+		--env "CC=x86_64-w64-mingw32-gcc" \
+		--rm \
+		$(BUILDER_IMAGE) \
+		$(1)
+endef
+
+# TODO: use `go list -f` to target the test files more directly
+ALL_GO_FILES = $(shell find . -name "*.go" -print | tr "\n" " ")
+GO_INTEG_TEST = go test -race -tags integration -c -o
+out/test-artifacts/linux-engine-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/engine)
+
+out/test-artifacts/linux-app-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/app)
+
+out/test-artifacts/linux-stats-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/stats)
+
+out/test-artifacts/windows-engine-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/engine)
+
+out/test-artifacts/windows-app-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/app)
+
+out/test-artifacts/windows-stats-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/stats)
+
+GO_FUNCTIONAL_TEST = go test -tags functional -c -o
+out/test-artifacts/linux-simple-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests/generated/simpletests_unix)
+
+out/test-artifacts/linux-handwritten-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests)
+
+out/test-artifacts/windows-simple-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests/generated/simpletests_windows)
+
+out/test-artifacts/windows-handwritten-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests)
+
+.PHONY: test-artifacts-windows test-artifacts-linux test-artifacts
+
+test-artifacts-windows: out/test-artifacts/windows-engine-tests.exe out/test-artifacts/windows-stats-tests.exe out/test-artifacts/windows-app-tests.exe out/test-artifacts/windows-simple-tests.exe out/test-artifacts/windows-handwritten-tests.exe
```

Comment:
I think you can break this into a few lines so it doesn't go so long: https://stackoverflow.com/questions/38086542/break-long-dependencies-in-makefile-into-several-lines